### PR TITLE
Rewrite the repetition penalty kernel for the larger maxSeqLen

### PR DESCRIPTION
**Pull request by StarrickLiu**
I have made improvements to the CUDA Kernel of TRTLLM by transitioning from a single thread block to multiple thread blocks, employing global memory allocated semaphores for inter-block synchronization.

企业微信截图_17032315883276
企业微信截图_17032316032542

Regarding precision, the modified kernel has successfully passed extensive unit tests. Randomly generated input data was used to feed both the new and old kernels for comparison. Across six thousand sets of data with varying BatchSize, VocabSize, SeqLen, and DataType, the output data from the new and old kernels were found to be completely identical.

企业微信截图_17032313924572

In terms of performance, I utilized Nsys Profile to run the aforementioned unit tests. Due to the use of semaphores for inter-block synchronization, the average execution time of the new kernel (33.551 us) increased by 18% compared to the old kernel (28.416 us). However, it theoretically supports an infinite SeqLen length (disregarding memory constraints).

Attachment:  
1、Code for unit tests.
[testRepetitionPenalty.zip](https://github.com/NVIDIA/TensorRT-LLM/files/13752080/testRepetitionPenalty.zip)
2、Nsys Profile Result
[Uploading report6.zip…]()

### CODE REVIEW

**Comment by byshiue** [comment]
Thanks for your support to TensorRT-LLM. This PR is not suitable for merging because it may cause a deadlock if seqBlockNum is too large (i.e., maxSeqLen is too large). We will soon add support for longer sequences. Thanks again for your interest in TensorRT-LLM.

**Comment by StarrickLiu** [comment]
> Thanks for your support to TensorRT-LLM. This PR is not suitable for merging because it may cause a deadlock if seqBlockNum is too large (i.e., maxSeqLen is too large). We will soon add support for longer sequences. Thanks again for your interest in TensorRT-LLM.

I see. I seem to have overlooked an issue in my code: since CUDA cannot guarantee a specific order of scheduling blocks, the only way to prevent such a deadlock is to limit the number of blocks in the grid.

Do you find the time overhead of duplicating logits before launching the kernel acceptable? Are there any other better approaches to implement a repetitionPenaltyKernel for long sequences?
